src/assets.rs

```diff
@@ -251,29 +251,42 @@ impl HighlightingAssets {
     }
 
     fn get_extension_syntax(&self, file_name: &OsStr) -> Option<&SyntaxReference> {
+        self.find_syntax_by_file_name(file_name).or_else(|| {
+            self.find_syntax_by_file_name_extension(file_name)
+                .or_else(|| self.get_extension_syntax_with_stripped_suffix(file_name))
+        })
+    }
+
+    fn find_syntax_by_file_name(&self, file_name: &OsStr) -> Option<&SyntaxReference> {
         self.get_syntax_set()
             .find_syntax_by_extension(file_name.to_str().unwrap_or_default())
-            .or_else(|| {
-                let file_path = Path::new(file_name);
-                self.get_syntax_set()
-                    .find_syntax_by_extension(
-                        file_path
-                            .extension()
-                            .and_then(|x| x.to_str())
-                            .unwrap_or_default(),
-                    )
-                    .or_else(|| {
-                        if let Some(file_str) = file_path.to_str() {
-                            for suffix in IGNORED_SUFFIXES.iter() {
-                                if let Some(stripped_filename) = file_str.strip_suffix(suffix) {
-                                    return self
-                                        .get_extension_syntax(OsStr::new(stripped_filename));
-                                }
-                            }
-                        }
-                        None
-                    })
-            })
+    }
+
+    fn find_syntax_by_file_name_extension(&self, file_name: &OsStr) -> Option<&SyntaxReference> {
+        let file_path = Path::new(file_name);
+        self.get_syntax_set().find_syntax_by_extension(
+            file_path
+                .extension()
+                .and_then(|x| x.to_str())
+                .unwrap_or_default(),
+        )
+    }
+
+    /// If we find an ignored suffix on the file name, e.g. '~', we strip it and
+    /// then try again to find a syntax without it. Note that we do this recursively.
+    fn get_extension_syntax_with_stripped_suffix(
```

Comment:
I know we have syntax highlighting test files to cover this, but maybe it could make sense to add some unit tests too, now that it is split out and thus should be easier to test?

Reply:
Thanks for the code review and suggestion!

I'm a big fan of end-to-end tests and integration tests, but not a super-big fan of unit tests. They certainly have their place in certain scenario, but I find that broad use of unit tests take long time to write and maintain, and tends to lock one in to a certain implementation since they are highly implementation dependent. In other words, they make refactoring hard.

I tried to be careful not to change any logic; only split things up into smaller methods, so given that our existing regression test pass, I'm confident enough in that I did not break anything to merge this.